File: C++_preparation/thisIsCodingTest/greedy/coin.cpp
#include <iostream>
#include <vector>

using namespace std;

int main()
{
    vector<int> coin = {500, 100, 50, 10};
    int cnt = 0;
    int money;
    cin >> money;
    for (int i = 0; i < coin.size(); i++)
    {
        cnt += money / coin[i];
        money %= coin[i];
    }
    cout << cnt << '\n';
    return 0;
}

File: codingTest_python/BOJ/coin.py
import sys
input = sys.stdin.readline

t = int(input())
for _ in range(t):
    N = int(input())
    coin = list(map(int, input().split()))
    money = int(input())
    dp = [0 for i in range(money+1)]
    dp[0] = 1
    for j in coin:
        for i in range(j, money+1):
            dp[i] += dp[i-j]

    print(dp[money])
